tools/fraken/main.go

```diff
@@ -0,0 +1,451 @@
+
+You may obtain a copy of the License at
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/hillu/go-yara/v4"
+)
+
+// Scanner is an instance of the Yara scanner.
+type Scanner struct {
+	rulesPath string
+	Rules     *yara.Rules
+}
+
+// Detection contains the information to report a Yara detection.
+type Detection struct {
+	ImagePath   string
+	SHA256      string
+	Signature   string
+	Description string
+	Reference   string
+	Score       int
+}
+
+var (
+	scanner           Scanner
+	scanPathFlag      *string
+	rulePathFlag      *string
+	magicPathFlag     *string
+	yaraRulesFlag     *string
+	magics            = make(map[string]string)
+	maxMagics         int
+	externalVariables = []string{"filepath", "filename", "filetype", "extension", "owner"}
+	MAX_GOROUTINES    = 10
+)
+
+func initMagics() {
+	// Try magic:
+	file, err := os.Open(*magicPathFlag) // For read access.
+	if err != nil {
+		// Try rule + magic
+		file, err = os.Open(path.Join(*rulePathFlag, *magicPathFlag))
+		if err != nil {
+			log.Printf("unable to open Magics file: %v\n", err)
+			return
+		}
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		t := scanner.Text()
+		if strings.HasPrefix(t, "#") || strings.TrimSpace(t) == "" {
+			continue
+		}
+		tokens := strings.Split(t, ";")
+		if len(tokens) != 2 {
+			log.Printf("unable to parse: %v", t)
+			continue
+		}
+		sig := strings.Replace(tokens[0], " ", "", -1)
+		if len(sig) > maxMagics {
+			maxMagics = len(sig)
+		}
+		magics[sig] = tokens[1]
+	}
+}
+
+func getFileTypes(filePath string) string {
+	f, err := os.Open(filePath)
+	if err != nil {
+		log.Printf("error opening file: %v\n", err)
+		return ""
+	}
+	defer f.Close()
+	fData := make([]byte, maxMagics)
+	f.Read(fData)
+	for k, v := range magics {
+		sig, err := hex.DecodeString(k)
+		if err != nil {
+			log.Printf("unable to parse signature %v: %v\n", k, err)
+			continue
+		}
+		if bytes.Equal(fData[:len(sig)], sig) {
+			return v
+		}
+	}
+	return "UNKNOWN"
+}
+
+func hashFile(filePath string) (string, error) {
+	h := sha256.New()
+	file, err := os.Open(filePath)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	_, err = io.Copy(h, file)
+	if err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(h.Sum(nil)), nil
+}
+
+func newDetection(imagePath, signature, description, reference string, score int) *Detection {
+	sha256, _ := hashFile(imagePath)
```

Comment:
I'd suggest not to ignore the error here and throw a warning to the user as it might indicate issues with reading the imagePath file

Reply:
Done